### PR TITLE
bugfix: fix dapp compat issues

### LINKED ISSUE #915: Ethfinex trustless not working
[Trustless Ethfinex](https://trustless.ethfinex.com/)  and [Trustless Ethfinex OTC](https://trustless.ethfinex.com/otc) are working fine on the extension but not working on Mobile

### LINKED ISSUE #951: EIP-712 chainID enforcement issues bug
chainId is an supposed to be optional parameter in the domain specification of eip-712

Currently chainId is not enforced at all by MetaMask mobile

Previously it was a requirement for any signTypedData_v3 request.

Request to have the functionality support enforcing chainId, but **ONLY** when chainId is present in the domain specification.

### PR DESCRIPTION
**Pull request by bitpshr**
**Description**

This pull request fixes dapp compatibility issues.

- `isApproved` should return `false` by default.
- `eth_personalSign` should ducktype address parameter.
- `eth_signTypedData_v3` should verify `chainId`.

**Checklist**

* [x] There is a related GitHub issue
* [ ] Tests are included if applicable
* [x] Any added code is fully documented

**Issue**

References https://github.com/MetaMask/metamask-mobile/issues/916
Resolves https://github.com/MetaMask/metamask-mobile/issues/915
Resolves https://github.com/MetaMask/metamask-mobile/issues/951

### CODE REVIEW

File: app/components/Views/BrowserTab/index.js [hunk 1]
```diff
@@ -61,6 +61,10 @@ import DrawerStatusTracker from '../../../core/DrawerStatusTracker';
 const { HOMEPAGE_URL } = AppConstants;
 const HOMEPAGE_HOST = 'home.metamask.io';
 
+function resemblesAddress(string) {
```
Comment (brunobar79) [inline app/components/Views/BrowserTab/index.js]: If we can't use `isValidAddress` from `ethereumjs-util` here, can you move it to app/util/address.js?

File: app/components/Views/BrowserTab/index.js [hunk 2]
```diff
@@ -507,6 +520,22 @@ export class BrowserTab extends PureComponent {
 			},
 			eth_signTypedData_v3: async payload => {
 				const { TypedMessageManager } = Engine.context;
+
+				let data;
+				try {
+					data = JSON.parse(payload.params[1]);
+				} catch (e) {
+					throw new Error('Data must be passed as a valid JSON string.');
```
Comment (brunobar79) [inline app/components/Views/BrowserTab/index.js]: I think if you throw here it won't get back to the dapp context.
You should do `Promise.reject({ error: error.message, jsonrpc: payload.jsonrpc, id: payload.id });`

File: app/components/Views/BrowserTab/index.js [hunk 3]
```diff
@@ -507,6 +520,22 @@ export class BrowserTab extends PureComponent {
...
+				}
+				const chainId = data.domain.chainId;
+				const activeChainId =
+					this.props.networkType === 'rpc' ? this.props.network : Networks[this.props.networkType].networkId;
+
+				// eslint-disable-next-line eqeqeq
+				if (chainId && chainId != activeChainId) {
+					throw new Error(`Provided chainId (${chainId}) must match the active chainId (${activeChainId})`);
```
Comment (brunobar79) [inline app/components/Views/BrowserTab/index.js]: same as previous comment

**Review by brunobar79 — APPROVED** [review]
LGTM,  added some minor comments

File: app/components/Views/BrowserTab/index.js (hunk 2 above)
Comment (bitpshr) [reply app/components/Views/BrowserTab/index.js]: Throwing here does get back to the dapp context (the thrown error is caught and shuffled as the RPC response up the chain.) I verified this only in a dapp context using Dan's test page, but I'm happy to explicitly reject if that's preferred.

File: app/components/Views/BrowserTab/index.js (hunk 1 above)
Comment (bitpshr) [reply app/components/Views/BrowserTab/index.js]: I wasn't sure if using `isValidAddress` was sufficient for all cases and defaulted to the logic the extension uses, which is where `resemblesAddress` came from. I will move this to the address util file for sure.

File: app/components/Views/BrowserTab/index.js (hunk 2 above)
Comment (bitpshr) [reply app/components/Views/BrowserTab/index.js]: I left this as-is for now to stay in-line with the way we handle invalid data above.

**Comment by ibrahimtaveras00** [comment]
There is an issue with LedgerDEX

When I navigate to [ LedgerDEX ](https://app.ledgerdex.com/#/) the page keeps refreshing and after signing to log in nothing happens; seen here = https://recordit.co/FtQBLcV1bv

**Pull request merged**